tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py

```diff
@@ -120,13 +120,79 @@ def tile_batch(t, multiplier, name=None):
     return nest.map_structure(lambda t_: _tile_batch(t_, multiplier), t)
 
 
-def _check_maybe(t):
+def _map_function(fn, t):
+  """Maps the function fn on each element of the TensorArray t."""
+  loop_stop = lambda mapped, elems, i: (
+      i == elems.size())
+  loop_body = lambda mapped, elems, i: (
+      mapped.write(i, fn(elems.read(i))), elems, i + 1)
+  mapped = tensor_array_ops.TensorArray(t.dtype, size=0, dynamic_size=True)
+  mapped, _, _ = control_flow_ops.while_loop(
```

Comment:
i have to think about this.  it sounds like at each time step you're reordering the entire history.  is that right?

Reply:
The idea is to keep the beams in the correct order at each timestep.

I guess it could also be done at the end of the decoding but the implementation here is more generic than just alignment history. The `TensorArray` might not contain time-based series.

Reply:
Thank you very much for the PR. I am also working on the same problem. Instead of picking out the right beams at each time step, we use gather_tree to sort out the right beams to avoid a quadratic complexity. Unfortunately the gather_tree only supports int32, so we did a dirty cast. It is fine for us as the alignments are only used for visualization on tensorboard. We update the finalize function and skip tensorarray for _merge_batch_beams and _split_batch_beams. 

```python
  def finalize(self, outputs, final_state, sequence_lengths):
    predicted_ids = beam_search_ops.gather_tree(
        outputs.predicted_ids, outputs.parent_ids,
        sequence_length=sequence_lengths)
    def gather_tree_from_array(t):
        if not isinstance(t, tensor_array_ops.TensorArray):
            return t
        history = t.stack()
        history = rnn._transpose_batch_time(history)
        history = math_ops.cast(history * 10000, dtypes.int32)
        history = beam_search_ops.gather_tree(
            history, outputs.parent_ids,
            sequence_length=sequence_lengths)
        history = math_ops.cast(history / 10000, t.dtype)
        return history
    final_state = nest.map_structure(gather_tree_from_array, final_state)
    outputs = FinalBeamSearchDecoderOutput(
        beam_search_decoder_output=outputs, predicted_ids=predicted_ids)
    return outputs, final_state
```